File: backend/app/models/vehiculo.py
"""
Modelo: Vehiculo.
Relación: Usuario 1:N Vehiculos.
"""

from sqlalchemy import String, Integer, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Vehiculo(Base):
    __tablename__ = "vehiculos"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    marca: Mapped[str] = mapped_column(String(100), nullable=False)
    modelo: Mapped[str] = mapped_column(String(100), nullable=False)
    anio: Mapped[int] = mapped_column(Integer, nullable=False)
    placa: Mapped[str] = mapped_column(
        String(20), unique=True, index=True, nullable=False
    )

    # FK
    usuario_id: Mapped[int] = mapped_column(
        ForeignKey("usuarios.id", ondelete="CASCADE"), nullable=False
    )

    # Relaciones
    propietario: Mapped["Usuario"] = relationship(back_populates="vehiculos")
    solicitudes: Mapped[list["SolicitudEmergencia"]] = relationship(
        back_populates="vehiculo"
    )

    def __repr__(self) -> str:
        return f"<Vehiculo(id={self.id}, placa='{self.placa}')>"

File: app_cliente/lib/features/vehiculos/models/vehiculo.dart
class Vehiculo {
  final int id;
  final String marca;
  final String modelo;
  final int anio;
  final String placa;

  Vehiculo({
    required this.id,
    required this.marca,
    required this.modelo,
    required this.anio,
    required this.placa,
  });

  factory Vehiculo.fromJson(Map<String, dynamic> json) {
    return Vehiculo(
      id: json['id'],
      marca: json['marca'],
      modelo: json['modelo'],
      anio: json['anio'],
      placa: json['placa'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'marca': marca,
      'modelo': modelo,
      'anio': anio,
      'placa': placa,
    };
  }
}
